### PR TITLE
Patch to allow building libpng on Macs with modern build system

### PR DESCRIPTION
**Pull request by slingthor**
### Description of Change(s)

Patches the building of libpng to work around an error when building with the modern build system.
Prevents the command generating pnglibconf.c from being attached to the following build targets:
genvers
gensym
genfiles

Since none of these targets are are common dependency of each other, this causes an error and is not allowed by the Xcode modern build system.

### Fixes Issue(s)
-Unable to build libpng on MacOS with modern build system

- [X] I have submitted a signed Contributor License Agreement

### CODE REVIEW

**Comment by meshula** [comment]
Please note that there is a fix here - https://github.com/glennrp/libpng/pull/403 - although it seems unlikely to land in a timely manner. Sadly the author of libpng passed away, and there aren't currently active maintainers (brief discussion here: https://github.com/glennrp/libpng/issues/410)

**Comment by slingthor** [comment]
> Please note that there is a fix here - [glennrp/libpng#403](https://github.com/glennrp/libpng/pull/403) - although it seems unlikely to land in a timely manner. Sadly the author of libpng passed away, and there aren't currently active maintainers (brief discussion here: [glennrp/libpng#410](https://github.com/glennrp/libpng/issues/410))

I agree the best way to work around it would be at the libpng level. Yea had heard about the author passing away, that's really sad.
Hopefully this patch can help until then. Also, is there a way to retrigger the test runner? I amended my commit and believe it may have confused it.

File: build_scripts/build_usd.py [hunk 1]
```diff
@@ -985,13 +985,20 @@ def InstallTIFF(context, force, buildArgs):
...
+                # Neon extensions
+                macArgs = ["-DPNG_HARDWARE_OPTIMIZATIONS=OFF",
+                          "-DPNG_ARM_NEON=off"] # case is significant
+            PatchFile("CMakeLists.txt",
+                [('add_custom_target(gensym DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/libpng.sym")',
+                  'add_custom_target(gensym DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/libpng.sym" genvers)'),
+                 ("add_custom_target(genfiles DEPENDS",
+                  "add_custom_target(genfiles DEPENDS gensym symbol-check")])
+
+        RunCMake(context, force, buildArgs)
```
Comment (davidgyu) [inline build_scripts/build_usd.py]: This is missing the "+ macArgs" that is needed to pass the Neon optimization args for Arm builds down to RunCMake().

File: build_scripts/build_usd.py [hunk 2]
```diff
@@ -985,13 +985,20 @@ def InstallTIFF(context, force, buildArgs):
...
+            if Arm():
+                # ensure libpng's build doesn't erroneously activate inappropriate
+                # Neon extensions
+                macArgs = ["-DPNG_HARDWARE_OPTIMIZATIONS=OFF",
+                          "-DPNG_ARM_NEON=off"] # case is significant
+            PatchFile("CMakeLists.txt",
+                [('add_custom_target(gensym DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/libpng.sym")',
+                  'add_custom_target(gensym DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/libpng.sym" genvers)'),
+                 ("add_custom_target(genfiles DEPENDS",
+                  "add_custom_target(genfiles DEPENDS gensym symbol-check")])
```
Comment (sunyab) [inline build_scripts/build_usd.py:1002]: Could a short comment be added here about why this patch is necessary? The build script tries to document why a patch is needed and what software versions are affected so we can understand them during future changes.

Something along the lines of, "This patch is needed to workaround a build issue on MacOS in libpng described at https://github.com/glennrp/libpng/issues/344. This was observed on CMake version blah / Xcode version blah" would be great. (Obviously feel free to add more information).

**Comment by slingthor** [comment]
Thanks for the valuable feedback so far guys! I've updated the PR to reflect the comments made!

**Comment by slingthor** [comment]
We've come to notice, that perhaps this is only happening for those who generate the project with --generator Xcode.
We can confirm, that if we use Jinja this does not occur.
The issue might be a full on blocker for generating with Xcode

**Comment by sunyab** [comment]
Filed as internal issue #USD-7489

**Comment by meshula** [comment]
@slingthor @davidgyu https://github.com/glennrp/libpng/pull/403 has landed!

**Comment by creijon** [comment]
> @slingthor @davidgyu [glennrp/libpng#403](https://github.com/glennrp/libpng/pull/403) has landed!

With the libPNG upgrade to v1.6.38 in https://github.com/PixarAnimationStudios/USD/pull/1959 this patch isn't required.

**Comment by slingthor** [comment]
Closed as no longer needed

**Pull request closed without merging**